jobs/import_delete/worker.go

```diff
@@ -0,0 +1,169 @@
+			return
+		case job := <-w.jobsChan:
+			mlog.Debug("Worker received a new candidate job.", mlog.String("worker", w.name))
+			w.doJob(&job)
+		}
+	}
+}
+
+func (w *ImportDeleteWorker) Stop() {
+	mlog.Debug("Worker stopping", mlog.String("worker", w.name))
+	close(w.stopChan)
+	<-w.stoppedChan
+}
+
+func (w *ImportDeleteWorker) doJob(job *model.Job) {
+	if claimed, err := w.jobServer.ClaimJob(job); err != nil {
+		mlog.Warn("Worker experienced an error while trying to claim job",
+			mlog.String("worker", w.name),
+			mlog.String("job_id", job.Id),
+			mlog.String("error", err.Error()))
+		return
+	} else if !claimed {
+		return
+	}
+
+	importPath := *w.app.Config().ImportSettings.Directory
+	retentionTime := time.Duration(*w.app.Config().ImportSettings.RetentionDays) * 24 * time.Hour
+	imports, appErr := w.app.ListDirectory(importPath)
+	if appErr != nil {
+		w.setJobError(job, appErr)
+		return
+	}
+
+	var hasErrs bool
+	for i := range imports {
+		filename := filepath.Base(imports[i])
+		modTime, appErr := w.app.FileModTime(filepath.Join(importPath, filename))
+		if appErr != nil {
+			mlog.Debug("Worker: Failed to get file modification time",
+				mlog.Err(appErr), mlog.String("import", imports[i]))
+			hasErrs = true
+			continue
+		}
+
+		if time.Now().After(modTime.Add(retentionTime)) {
+			// expected format if uploaded through the API is
+			// ${uploadID}_${filename}${app.IncompleteUploadSuffix}
+			minLen := 26 + 1 + len(app.IncompleteUploadSuffix)
+
+			// check if it's an incomplete upload and attempt to delete its session.
+			if len(filename) > minLen && filepath.Ext(filename) == app.IncompleteUploadSuffix {
+				uploadID := filename[:26]
+				if storeErr := w.app.Srv().Store.UploadSession().Delete(uploadID); storeErr != nil {
+					mlog.Debug("Worker: Failed to delete UploadSession",
+						mlog.Err(storeErr), mlog.String("upload_id", uploadID))
+					hasErrs = true
+				}
+			} else {
+				// check if fileinfo exists and if so delete it.
+				filePath := filepath.Join(imports[i])
+				info, storeErr := w.app.Srv().Store.FileInfo().GetByPath(filePath)
+				var nfErr *store.ErrNotFound
+				if storeErr != nil && !errors.As(storeErr, &nfErr) {
+					mlog.Debug("Worker: Failed to get FileInfo",
+						mlog.Err(storeErr), mlog.String("path", filePath))
+					hasErrs = true
+				} else if storeErr == nil {
+					if storeErr = w.app.Srv().Store.FileInfo().PermanentDelete(info.Id); storeErr != nil {
+						mlog.Debug("Worker: Failed to delete FileInfo",
+							mlog.Err(storeErr), mlog.String("file_id", info.Id))
+						hasErrs = true
+					}
+				}
+			}
+
+			// remove file data from storage.
+			if appErr := w.app.RemoveFile(imports[i]); appErr != nil {
+				mlog.Debug("Worker: Failed to remove file",
+					mlog.Err(appErr), mlog.String("import", imports[i]))
+				hasErrs = true
+				continue
+			}
+		}
+	}
+
+	if hasErrs {
+		mlog.Warn("Worker: errors occurred")
```

Comment:
I think we should `w.setJobError` here

Reply:
I suppose we could. The only issue is that the import files could have been deleted by then. I opted for not failing fast although that could also be an option, to exit the loop on first error.

Reply:
I see your point. My concern is getting an import to an inconsistent state where a deletion retry will not catch the sate and try to delete, and the only one I can think of is one in which we delete the files, but not the DB entities related.

So what about instead of marking the job as failed, checking `if hasErrs` before running the actual file deletion? That way the files will only be deleted if the upload session or file info has been completely cleared first, and if they caused any errors, the import files will be there for the next iteration and they will be selected again for deletion.

Reply:
We don't really care about DB as much. Our primary state is the filesystem since we support externally uploaded imports. If a DB action fails the worst it can happen is that we delete the file and we are left with a FileInfo/UploadSession entry pointing to a non-existent path. That said we can add a check for consistency sake, it's not a big deal :)

Reply:
The other issue with `hasErrs` is that it's global, not per file, so we would be skipping every delete file operation after first setting it. If we do that we might as well just fail the job on first error and exit. Unless we introduce another variable to keep track of errors for a specific file but that might make things overcomplicated.

Reply:
Yeah, I was thinking on a local `hasErrs`, but if we don't care that much about a `FileInfo` pointing nowhere, I agree, let's not overcomplicate this. If this proves to be something to improve on, the changes would not be big, so I'm happy to merge it as is and go back if needed

Reply:
I think I was able to obtain what you suggested without too much logic change, we simply continue to the next file if some error occurs. Let me know if that works.

Reply:
Right, I think it works perfect!